1-Wire API, with zephyr-serial driver and DS18B20 Sensor

str4t0m opened the pull request:
This PR introduces an API for 1-Wire devices.
Additionally, the zephyr-serial driver implements this API using serial polling operations.
More optimized implementations could be added later, but a driver on top of serial polling should work on most platforms.
Basic API tests have been added to make sure the driver compiles without errors.
In order to demonstrate the API and bindings for 1-Wire, also a DS18b20 driver and a sample are added.
The API mostly matches the one proposed in #6875 with a few modifications and additions.

Closes #32516.

str4t0m committed:
CODEOWNERS: MAINTAINERS: Add entry for w1 support

Add entries for w1 doc, api, drivers, and tests in the CODEOWNERS as well
as the MAINTAINERS file.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
tests: drivers: w1: add basic api test

This commit adds basic api tests for the w1 driver.

These tests mostly make sure that everything compiles without
errors. They expect a w1-bus configuration without any connected slave.
No bidirectional read/write tests are executed.
Send tests simply send without verifying the output,
and receive tests expect an idle bus(receiving only logical ones).

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
drivers: sensor: add support for ds18b20 1-wire temperatue sensor

This commit adds support for the ds18b20 1-wire temperature sensor.

The sampling resolution of the sensor can be set in DT.
In case only a single device is on the bus, the driver issues
skip_rom commands. However, in case DT defines several devices,
the driver will use match_rom commands and therefore it is necessary
to set the rom_id of the device via the sensor attribute interface before
being able to sample sensor values.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
samples: sensor: ds18b20: add sample for this temperature driver

This commit adds a sample for the maxim ds18b20 1-wire temperatue
sensor in a single-drop bus configuration.
Includes overlay files for nucleo_g0b1re und nrf52840dk_nrf52840 with
the required open drain configuration.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
tests: drivers: build_all: sensor: add w1 driver and ds18b20

This commit adds the necessary changes to build w1-devices using
the vnd,w1 dummy driver in the build_all test case.

Additionally, definitions for the ds18b20 1-wire temperature sensor
are added.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
drivers: w1: add zephyr-serial driver

The zephyr-serial w1 driver introduced in this commit implements
all routines for the w1 api on top of the zephyr serial driver.
W1 bit read, write, and reset operations are executed by issuing
polling zephyr serial byte read and write operations.

The driver should be usable on most platforms in zephyr that have
implemented support for the polling procedures of the serial driver.
As not all serial drivers are implemented exactly the same minor
additional quirks may be needed on some platforms.
The most notable difference of polling serial driver implementations
seems to be that some return immediately from poll_out after the
transmission was started(e.g. STM32) and others wait until
the transmission was completed before returning from poll_out
(e.g. NRF). While this has influence on the timeout, both types
are supported by this driver because the driver waits for a
configurable time period until it terminates the read.

The driver needs an appropriate open drain interface to be able
to communicate with slaves.
In the simpliest case this might be achived by configuring the mcu pins
in open-drain configuration with a (sufficiently small) pull-up to 3V3/5V.
Otherwise an external circuit needs to provide this interface.

Overdrive and Standard Speed modes are supported by this driver.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
doc: add api documentation for w1 driver

This commit adds the w1 driver API documentation in peripherals section
of the reference guide.
The 1-Wire API is declared as unstable in the API overview.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

tagunil commented on drivers/w1/w1_zephyr_serial.c:
```diff
@@ -0,0 +1,265 @@
+	return rx_bit == W1_SERIAL_READ_REQ_BYTE;
+}
+
+void  w1_serial_write_bit(const struct device *dev, const bool bit)
```
```suggestion
void w1_serial_write_bit(const struct device *dev, const bool bit)
```

str4t0m replied:
Fixed

tagunil commented on drivers/w1/w1_zephyr_serial.c:
```diff
@@ -0,0 +1,265 @@
+	return rx_byte;
+}
+
+void  w1_serial_write_byte(const struct device *dev, const uint8_t byte)
```
```suggestion
void w1_serial_write_byte(const struct device *dev, const uint8_t byte)
```

carlescufi commented:
@str4t0m do you have plans to move this forward? We can bring the proposed API to our regularly scheduled API meeting for discussion.

MarcianoPreciado commented:
This is a pretty comprehensive first addition of a 1-Wire interface! Thanks for putting this together!

str4t0m commented:
Hi,
recently I have been a bit busy with other stuff, therefore I have not done any upstream contributions.
But if there is interest, I will find some time to move this PR forward.
There were a few things I have not been completely happy, when I opened the PR initially therefore I have left it as a draft for now. For example getting the rom-id from devicetree does not really make sense as it is unique per device. I will rebase and review the PR until the end of the week and welcome any feedback.
I chose an implementation on a simple serial driver as this was sufficient for my project. Additionally using the serial interface we don't need any timing tests(yet) and should still support most SoCs. Therefore I thought this should be sufficient to get started the discussion about the API and add bit-banged or more optimized implementations later.

de-nordic commented:
@str4t0m The support code looks nice. I have some comments regarding structure:
I think that the support for 1-Wire device, in form of API should be divided into three layers:
1) physical: reset/detect, single bit tx/rx, single byte tx/rx, line level (to check for example if line is kept low);
1a) multi-byte rx and tx and combined tx/rx (something like send one byte of command get N back);
2) Utils, generic protocol specific: scan for devices, get id, skip rom, match rom, and so on; you know the common codes;
3) Drivers for specific devices - these will require specific API, for example to get temperature or to be attached to existing API, in case of EEPROM devices.

The 1-2 should be exposed to users as API, so that a 1-Wire device could be used without driver, directly from software.
The 1) has been divided into 1) and 1a) because multi-byte, or combined operations, would be generally implemented in software; having 1) as separate API allows to add support for devices like the DS2482-100 i2c-1w bridge (https://www.maximintegrated.com/en/products/interface/controllers-expanders/DS2482-100.html#tech-docs), using the same API as the uart or bit-banging implementation.

str4t0m committed:
drivers: sensor: extend with w1 api for ROM config

This commit adds a sensor attribute and helper functions such that
the unique ROM of 1-Wire devices can easily configured.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m commented:
@de-nordic thanks for the feedback and sorry for the long delay.
I have now rebased and have made several smaller modifications, most notably: 
- add w1_resume_command
- add w1_write_read
- remove Devicetree ROM ID configuration
- add helpers to be able to set the ROM ID of Sensors via the attribute interface.

Regarding the structure. Your suggestions regarding the layered API seem reasonable to me, I believe I have roughly implemented it already according to those layers, but I definitely have not separated the layers clearly enough.
The layer 1) would mostly correspond to the "link layer" of the specification and 2) would correspond to the "network layer".

For me it is not yet completely clear how you have envisioned the split into 1) and 1a).
Would 1a) simply call 1) functions, or would you allow different implementations for 1a) for devices that support multi-byte 1-Wire r/w operations like DS2485 I2c to1-Wire master. I once already considered replacing the multi-byte read and write functions from the hardware dependent API with generic ones that simply call the layer 1) single-byte rw operations multiple times, but then I realized that there exist several SPI/I2c to 1-Wire bridges(like DS2485) that support multi byte writes. I have never used such external 1-Wire Interfaces and don't know how commonly they are used, but, as you have pointed out above, we should consider those in the design of the API.
Looking into some datasheets, a so called "triplet command" is often available on those devices. The triplet command is used for search operations in the device discovery process. Should we extend the layer 1)  for such a command, or should we leave this out for now. Some external masters can even offload the complete search and crc calculation, but I think those are not so important.
Inspired by your suggestion of a combined rx/tx command I have now added a w1_write_read command. I have added it to layer 2) because it additionally executes a reset_select.

link layer (1) - network layer(2) split: Does the current status approximately match your expectations, or would you split those completely in different header files, ... ?

overdrive mode: The overdrive mode is currently not supported, should we already discuss how the API should be extended to support the overdrive mode or could this be left as a possible future addition.

Strong pullup: Some 1-wire devices have power requirements that exceed the maximal power which can be drawn over the pull-up resistor. Therefore they require the master to activate a strong pull-up after a specified set of commands. The currently proposed API does not provide any means for such actions. Strong pull ups are required by several EEPROMs and devices with an integrated crypto accelerator like the DS28E40 Authenticator.

Line level: I have not yet extended the 1) layer. Would you use this to check if a hw or sw fault may keep the line low?
I suppose the read bit command is not sufficient for your use case? In that case the serial driver should likely also avoid sending any start and stop bit during the process of this measurement and instead use the gpio driver to read the level?

Locking: Currently all network functions lock the bus to make sure the operations are not interrupted. Other zephyr drivers usually don't do this. Could you also weight in if we should transfer this responsibility to the user?

Search: The current implementation searches the bus for devices and executes a callback for each found device. Alternatives I have considered would be to 1) execute the callback only once returning all found devices and 2) instead provide two methods "start search" and "search next". The second one is used in several libraries because it's very similar to what some external 1-wire masters provide commands for.

Terminology: As far as I have understood Zephyr tries to avoid oppressive language, but generally keeps using well known identifiers until new terms are officially approved by the respective standards organization. The 1-Wire specification refers to the bus participants using the widely distributed master/slave terminology, and I don't expect Maxim to release any new documentations where they change this soon. Therefore I am curious if I should try to avoid using the master slave terminology or if doing so would only needlessly complicate the 1-Wire drivers.

branilson commented:
Hello, @str4t0m.

I following this PR for some time and see you are doing a good job, but it still has a pending test to pass. Is it possible to use this lib out of the stack? How can I do it?

Greetings.

ycsin commented:
> Is it possible to use this lib out of the stack? How can I do it?

You can cherry pick the commit into your branch and initialize the zephyr base with your branch instead of zephyr upstream:

```
west init -m https://github.com/branilson/zephyr --mr your_branch zephyrproject
```

dp7hgh7 commented:
Hi,
I'm currently implementing a driver for the DS2484 from Maxim which is an I2C device. I would like to use this api but right now it's impossible to tell the application whether the call failed on the I2C line or the 1-wire operation had no result. E. g.
```c
bool w1_reset_bus(const struct device *dev);
```
should return whether a device is present on the bus (true) or not (false). This applies to most of the functions. Is there a reason not to do it like
```c
int w1_reset_bus(const struct device *dev, bool *devices_present);
```
?

Besides that: Great work!

Regards

dp7hgh7 commented on drivers/sensor/ds18b20/ds18b20.c:
```diff
@@ -0,0 +1,256 @@
+		return -ENOTSUP;
+	}
+
+	return 0;
```
Should return `rc`, right? Otherwise it would return 0 even if `rc` is set to `-ENOTSUP`.
```suggestion
	return rc;
```

str4t0m replied:
Fixed.

str4t0m committed:
drivers: w1: add vnd,w1 driver for ci testing

This commit adds the vnd,w1 dummy host driver used for ci builds of
1-wire devices.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
drivers: w1 introduce 1-wire api and common code

This commit introduces a new api for the Dallas 1-wire protocol.
The api includes link functions for read and write operations on
bit, byte, and block level, as well as functions to reset and
lock the bus.

The bus configuration is derived from the device tree and can be
queried using w1_slave_count routine.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

de-nordic commented on drivers/sensor/ds18b20/ds18b20.c:
```diff
@@ -0,0 +1,268 @@
+		w1_sensor_value_to_rom(thr, &data->config.rom_id);
+	};
+
+	return rc;
```
This may return uninitialized rc.

str4t0m replied:
Fixed.

de-nordic commented on drivers/sensor/ds18b20/ds18b20.c:
```diff
@@ -0,0 +1,268 @@
+		data->lazy_loaded = true;
+	}
+
+	ret = ds18b20_temperature_convert(dev);
+	if (ret < 0) {
+		LOG_DBG("W1 fetch error");
+		return -EIO;
+	}
```
Since `ret` is not returned anywhere it does not seem to be needed; there is also `status` variable that shares the same purpose.

str4t0m replied:
I changed it to assign the value to status, and return it afterwards.

de-nordic commented on include/drivers/w1/w1_net.h:
```diff
@@ -0,0 +1,316 @@
+ *         on error.
+ */
+int w1_reset_select(const struct device *dev, const struct w1_config *config);
+/* int w1_reset_select_od(const struct device *dev, const struct w1_rom *rom_id); */
```
Some leftover?

str4t0m replied:
Yes

de-nordic commented on include/drivers/w1/w1_net.h:
```diff
@@ -0,0 +1,316 @@
+ * @retval              Number of devices found.
+ * @retval              -EIO if search failed due to w1 read/write issues.
+ */
+/* TODO need config? */
```
?

de-nordic commented on drivers/w1/w1_net_intern.c:
```diff
@@ -0,0 +1,203 @@
+			 */
+		} else {
+			/*
+			 * (!last_id_bit && !last_complement_id_bit)
```
?

str4t0m replied:
This comments only the the condition for taking this branch, so it is really not needed. I will remove it.

str4t0m commented on include/drivers/w1/w1.h:
```diff
@@ -0,0 +1,358 @@
+
+	/* if ((flags & W1_SPEED_OVERDRIVE) && !ctrl_data->overdrive_active) { */
+
+	/* } */
```
Yes, indeed.

carlescufi commented on drivers/w1/Kconfig:
```diff
@@ -0,0 +1,33 @@
+rsource "Kconfig.zephyr_serial"
+
+config W1_NET
+	bool "1-Wire network layer"
```
`default y`

str4t0m replied:
Done

str4t0m commented on drivers/sensor/ds18b20/ds18b20.h:
```diff
@@ -0,0 +1,66 @@
+/* convert resolution in bits to scratchpad config format */
+#define DS18B20_RESOLUTION(res)		((res - 9) << 5)
+/* convert resolution in bits to array index (for resolution specific elements) */
+#define DS18B20_RESOLUTION_INDEX(res)	((res - 9))
```
I am now using the BIT_MASK macro for the first, but this may not be how you have envisioned it?

str4t0m commented:
As discussed, the headers are now merged into a single file.
Also, all the common code is now merged into `drivers/w1/w1_net.c`
I have not added the w1_triplet() to the low level APi, as I believe it should be introduced later together with a driver that actually implements it.

ycsin commented on doc/reference/peripherals/w1.rst:
```diff
@@ -0,0 +1,94 @@
+In Zephyr this API is split into the following layers.
+
+* The link layer handles basic communication functions such as bus reset,
+  present detect and bit transfer operations.
```
presence detection?

ycsin replied:
```suggestion
  presence detection and bit transfer operations.
```

ycsin commented on doc/reference/peripherals/w1.rst:
```diff
@@ -0,0 +1,94 @@
+
+* The link layer handles basic communication functions such as bus reset,
+  present detect and bit transfer operations.
+  It is the only hardware dependent layer in Zephyr.
```
hardware-dependent

ycsin replied:
```suggestion
  It is the only hardware-dependent layer in Zephyr.
```

ycsin commented on doc/reference/peripherals/w1.rst:
```diff
@@ -0,0 +1,94 @@
+  present detect and bit transfer operations.
+  It is the only hardware dependent layer in Zephyr.
+  This layer is supported by a driver using the zephyr-serial interface, which should
+  work on most zephyr platforms. In the future a GPIO/Timer based driver and hardware
```
```suggestion
  work on most zephyr platforms. In the future, a GPIO/Timer based driver and hardware
```

ycsin commented on doc/reference/peripherals/w1.rst:
```diff
@@ -0,0 +1,94 @@
+
+   * All slave devices have a unique 64-bit identification number, which includes a
+     8-bit `1-Wire Family Code`_ and a 8-bit CRC.
+   * In order to find devices on the bus the standard specifies an search algorithm
```
```suggestion
   * In order to find devices on the bus, the standard specifies an search algorithm
```

ycsin commented on drivers/sensor/ds18b20/ds18b20.c:
```diff
@@ -0,0 +1,267 @@
+static int ds18b20_write_scratchpad(const struct device *dev,
+				    struct ds18b20_scratchpad scratchpad)
+{
+
```
```suggestion
```

ycsin commented on drivers/sensor/ds18b20/ds18b20.c:
```diff
@@ -0,0 +1,267 @@
+{
+	struct ds18b20_data *data = dev->data;
+	const struct device *bus = ds18b20_bus(dev);
+	uint8_t cmd =  DS18B20_CMD_READ_SCRATCHPAD;
```
```suggestion
	uint8_t cmd = DS18B20_CMD_READ_SCRATCHPAD;
```

ycsin commented on drivers/sensor/ds18b20/ds18b20.c:
```diff
@@ -0,0 +1,267 @@
+ * Write resolution into configuration struct,
+ * but don't write it to the sensor yet.
+ */
+static int ds18b20_set_resolution(const struct device *dev, uint8_t resolution)
```
Would it be better if this is `void` instead?

ycsin commented on drivers/sensor/ds18b20/ds18b20.c:
```diff
@@ -0,0 +1,267 @@
+		LOG_DBG("W1 fetch error");
+		return status;
+	}
+	k_sleep(K_MSEC(measure_wait_ms[
+				DS18B20_RESOLUTION_INDEX(cfg->resolution)]));
```
```suggestion
	k_msleep(measure_wait_ms[DS18B20_RESOLUTION_INDEX(cfg->resolution)]);
```

ycsin commented on drivers/sensor/ds18b20/ds18b20.c:
```diff
@@ -0,0 +1,267 @@
+{
+	struct ds18b20_data *data = dev->data;
+
+	if (chan == SENSOR_CHAN_AMBIENT_TEMP) {
+		ds18b20_temperature_from_raw((uint8_t *)&data->scratchpad.temp,
+					     val);
+	} else {
+		return -ENOTSUP;
+	}
+
+	return 0;
```
```suggestion
	if (chan != SENSOR_CHAN_AMBIENT_TEMP) {
		return -ENOTSUP;
	}

	ds18b20_temperature_from_raw((uint8_t *)&data->scratchpad.temp, val);
	return 0;
```

ycsin commented on drivers/sensor/ds18b20/ds18b20.c:
```diff
@@ -0,0 +1,267 @@
+		w1_sensor_value_to_rom(thr, &data->config.rom_id);
+	};
+
+	return rc;
```
I dont see the need of having `rc` here tbh, you could just return `-ENOTSUP` or `0` directly.

ycsin commented on drivers/w1/w1_net.c:
```diff
@@ -0,0 +1,405 @@
+ * W1_SEARCH_LAST_DEVICE, and in case no device participated in the search,
+ * the discrepancy is set to W1_SEARCH_NO_DEVICE.
+ *
+ * The Implementation is similar as suggested in the maxim application note 187.
```
```suggestion
 * The implementation is similar as suggested in the maxim application note 187.
```

ycsin commented on drivers/w1/w1_net.c:
```diff
@@ -0,0 +1,405 @@
+	}
+	next_discrepancy = W1_SEARCH_LAST_DEVICE;
+
+	for (size_t id_bit_nr = 1; id_bit_nr < 65; id_bit_nr++) {
```
Probably looks better if 1 & 65 is defined as a macro

ycsin replied:
`W1_SEARCH_LAST_DEVICE`?

str4t0m replied:
Now using W1_SEARCH_LAST_DEVICE as limit, but left 1 as start value because I am not sure if introducing a macro here makes it clearer.

ycsin replied:
I think its good enough

ycsin commented on drivers/w1/w1_net.c:
```diff
@@ -0,0 +1,405 @@
+
+		ret = w1_read_bit(dev);
+		if (ret < 0) {
+			return -EIO;
```
```suggestion
			return ret;
```

ycsin commented on drivers/w1/w1_net.c:
```diff
@@ -0,0 +1,405 @@
+		last_id_bit = (bool)ret;
+		ret = w1_read_bit(dev);
+		if (ret < 0) {
+			return -EIO;
```
```suggestion
			return ret;
```

ycsin commented on drivers/w1/w1_net.c:
```diff
@@ -0,0 +1,405 @@
+		 */
+		ret = w1_write_bit(dev, last_id_bit);
+		if (ret < 0) {
+			return -EIO;
```
```suggestion
			return ret;
```

ycsin commented on drivers/w1/w1_net.c:
```diff
@@ -0,0 +1,405 @@
+				    &found_rom_inv_64);
+		if (ret < 0) {
+			found_cnt = ret;
+			goto out;
```
break?

ycsin commented on drivers/w1/w1_net.c:
```diff
@@ -0,0 +1,405 @@
+		 * Only big-endian targets need to swap, such that struct's
+		 * bytes are stored in big-endian byte order.
+		 */
+#if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+		sys_memcpy_swap(&found_rom, &found_rom_inv_64, 8);
+#else
+		*(uint64_t *)&found_rom = found_rom_inv_64;
+#endif
```
For the sake of readability, I might do it like this instead:
```suggestion
		if (__BYTE_ORDER__ == __ORDER_BIG_ENDIAN__) {
			sys_memcpy_swap(&found_rom, &found_rom_inv_64, 8);
		} else {
			*(uint64_t *)&found_rom = found_rom_inv_64;
		}
```

ycsin commented on drivers/w1/w1_net.c:
```diff
@@ -0,0 +1,405 @@
+
+int w1_read_rom(const struct device *dev, struct w1_rom *rom_id)
+{
+	int ret = 0;
```
```suggestion
	int ret;
```

ycsin commented on drivers/w1/w1_net.c:
```diff
@@ -0,0 +1,405 @@
+
+static int match_rom(const struct device *dev, const struct w1_config *config)
+{
+	int ret = 0;
```
```suggestion
	int ret;
```

ycsin commented on drivers/w1/w1_net.c:
```diff
@@ -0,0 +1,405 @@
+
+int w1_match_rom(const struct device *dev, const struct w1_config *config)
+{
+	int ret = 0;
```
```suggestion
	int ret;
```

ycsin commented on drivers/w1/w1_net.c:
```diff
@@ -0,0 +1,405 @@
+
+int w1_resume_command(const struct device *dev)
+{
+	int ret = 0;
```
```suggestion
	int ret;
```

ycsin commented on drivers/w1/w1_net.c:
```diff
@@ -0,0 +1,405 @@
+
+static int skip_rom(const struct device *dev, const struct w1_config *config)
+{
+	int ret = 0;
```
```suggestion
	int ret;
```

ycsin commented on drivers/w1/w1_net.c:
```diff
@@ -0,0 +1,405 @@
+	int ret = 0;
+
+	w1_lock_bus(dev);
+	ret = w1_reset_bus(dev);
+	if (ret == 0) {
+		ret = -ENODEV;
```
`ENODEV` seems reasonable

ycsin commented on drivers/w1/w1_net.c:
```diff
@@ -0,0 +1,405 @@
+
+int w1_skip_rom(const struct device *dev, const struct w1_config *config)
+{
+	int ret = 0;
```
```suggestion
	int ret;
```

ycsin commented on drivers/w1/w1_net.c:
```diff
@@ -0,0 +1,405 @@
+
+static int reset_select(const struct device *dev, const struct w1_config *config)
+{
+	int ret = 0;
```
```suggestion
	int ret;
```

ycsin commented on drivers/w1/w1_net.c:
```diff
@@ -0,0 +1,405 @@
+
+static int reset_select(const struct device *dev, const struct w1_config *config)
+{
+	int ret = 0;
+
+	if (w1_get_peripheral_count(dev) > 1) {
+		ret = match_rom(dev, config);
+	} else {
+		ret = skip_rom(dev, config);
+	}
+
+	return ret;
```
```suggestion
	if (w1_get_peripheral_count(dev) > 1) {
		return match_rom(dev, config);
	}

	return skip_rom(dev, config);

```

ycsin commented on drivers/w1/w1_net.c:
```diff
@@ -0,0 +1,405 @@
+
+int w1_reset_select(const struct device *dev, const struct w1_config *config)
+{
+	int ret = 0;
```
```suggestion
	int ret;
```

ycsin commented on drivers/w1/w1_net.c:
```diff
@@ -0,0 +1,405 @@
+		      const uint8_t *write_buf, size_t write_len,
+		      uint8_t *read_buf, size_t read_len)
+{
+	int ret = 0;
```
```suggestion
	int ret;
```

ycsin commented on drivers/w1/w1_net.c:
```diff
@@ -0,0 +1,405 @@
+		  const uint8_t *write_buf, size_t write_len,
+		  uint8_t *read_buf, size_t read_len)
+{
+	int ret = 0;
```
```suggestion
	int ret;
```

ycsin commented on drivers/w1/w1_test.c:
```diff
@@ -0,0 +1,81 @@
+}
+
+static const struct w1_driver_api w1_vnd_api = {
+	.reset_bus		= w1_vnd_reset_bus,
+	.read_bit		= w1_vnd_read_bit,
+	.write_bit		= w1_vnd_write_bit,
+	.read_byte		= w1_vnd_read_byte,
+	.write_byte		= w1_vnd_write_byte,
+	.configure		= w1_vnd_configure,
```
I think this isn't inline with the Zephyr's clang-format

ycsin commented on drivers/w1/w1_zephyr_serial.c:
```diff
@@ -0,0 +1,289 @@
+}
+
+/* Concurretly tranmits and receives one 1-Wire byte */
+static uint8_t serial_tx_rx_byte(const struct device *dev, uint8_t tx_byte,
```
```suggestion
static int serial_tx_rx_byte(const struct device *dev, uint8_t tx_byte,
```

ycsin commented on dts/bindings/w1/w1-device.yaml:
```diff
@@ -0,0 +1,24 @@
+    overdrive-speed:
+      type: boolean
+      description: |
+          Instead of standard speed timing use overdrive speed timing for
```
```suggestion
          Instead of standard speed timing, use overdrive speed timing for
```

ycsin commented on include/drivers/w1.h:
```diff
@@ -0,0 +1,647 @@
+{
+	const struct w1_driver_api *api = dev->api;
+
+	return api->configure(dev, type, value);
+
+	return 0;
```
```suggestion
	return api->configure(dev, type, value);
```

ycsin commented on samples/sensor/ds18b20/prj.conf:
```diff
@@ -0,0 +1,10 @@
+CONFIG_SENSOR=y
+CONFIG_W1=y
+CONFIG_W1_NET=y
```
This should be default to `y` already in the Kconfig?

ycsin commented on tests/drivers/build_all/sensor/prj.conf:
```diff
@@ -8,6 +8,8 @@ CONFIG_I2C=y
 CONFIG_SPI=y
 CONFIG_SENSOR=y
+CONFIG_W1=y
+CONFIG_W1_NET=y
```
This should be default to `y` already in the Kconfig?

ycsin commented on tests/drivers/w1/w1_api/prj.conf:
```diff
@@ -0,0 +1,4 @@
+CONFIG_ZTEST=y
+CONFIG_W1=y
+CONFIG_W1_NET=y
```
This should be default to `y` already in the Kconfig?

ycsin reviewed (changes requested):
Thanks for this @str4t0m 
Looks reasonable overall, I'd probably cast the unhandled functions to `void` and clang-format everything.

str4t0m commented:
@ycsin Thanks for your review. I have addressed your comments now.

ycsin commented on drivers/sensor/ds18b20/ds18b20.c:
```diff
@@ -0,0 +1,257 @@
+	return w1_write_read(bus, &data->config, sp_data, sizeof(sp_data), NULL, 0);
+}
+
+static int ds18b20_read_scratchpad(const struct device *dev,
```
@str4t0m isn't it `int`?

str4t0m replied:
I am not sure what you mean here. Do you want to say that I should check the return value of each ds18b20_read_scratchpad() call?

ycsin replied:
I think the practice is to return unhandled error to the caller wherever possible, so I'm just suggesting that maybe you can return `ds18b20_read_scratchpad(dev, &data->scratchpad);` instead of 0

str4t0m replied:
Yes this make sense. I am now returning the status in line 140.

ycsin commented on drivers/sensor/ds18b20/ds18b20.c:
```diff
@@ -0,0 +1,257 @@
+	return 0;
+}
+
+int ds18b20_attr_set(const struct device *dev, enum sensor_channel chan,
+		     enum sensor_attribute attr, const struct sensor_value *thr)
+{
+	struct ds18b20_data *data = dev->data;
+
+	if (chan != SENSOR_CHAN_ALL && chan != SENSOR_CHAN_AMBIENT_TEMP) {
+	} else if ((enum sensor_attribute_w1)attr == SENSOR_ATTR_W1_ROM_ID) {
+		data->lazy_loaded = false;
+		w1_sensor_value_to_rom(thr, &data->config.rom_id);
+		return 0;
+	};
+
+	return -ENOTSUP;
+}
```
```suggestion
int ds18b20_attr_set(const struct device *dev, enum sensor_channel chan,
		     enum sensor_attribute attr, const struct sensor_value *thr)
{
	struct ds18b20_data *data = dev->data;

	if ((enum sensor_attribute_w1)attr == SENSOR_ATTR_W1_ROM_ID) {
		data->lazy_loaded = false;
		w1_sensor_value_to_rom(thr, &data->config.rom_id);
		return 0;
	}

	return -ENOTSUP;
}
```

str4t0m replied:
I do Agree, checking the channel is not really necessary for this attribute.

ycsin replied:
I simply 'factorized` this function.

It doesn't do anything when `(chan != SENSOR_CHAN_ALL && chan != SENSOR_CHAN_AMBIENT_TEMP)` and return `-ENOTSUP`, and it returns 0 only when `((enum sensor_attribute_w1)attr == SENSOR_ATTR_W1_ROM_ID)`

ycsin replied:
You could do it like this as well:

```c
	struct ds18b20_data *data = dev->data;

	if ((enum sensor_attribute_w1)attr != SENSOR_ATTR_W1_ROM_ID) {
		return -ENOTSUP;
	}

	data->lazy_loaded = false;
	w1_sensor_value_to_rom(thr, &data->config.rom_id);
	return 0;
	
```

str4t0m commented on doc/reference/peripherals/w1.rst:
```diff
@@ -0,0 +1,94 @@
+1-Wire is a low speed half-duplex serial bus using only a single wire plus
+ground for both data transmission and device power supply.
+Similarly to I2C, 1-Wire uses a bidirectional open-collector data line,
+is a single master multidrop bus, where one controller (master) initiates all
```
I have reformulated this sentence.

ycsin commented:
This PR looks pretty good to me, and would be a significant addition to Zephyr.
@carlescufi how can we move this PR forward?

ycsin reviewed (dismissed):
This PR looks pretty good to me, and would be a significant addition to Zephyr.
@carlescufi how can we move this PR forward?

de-nordic commented on drivers/sensor/ds18b20/ds18b20.c:
```diff
@@ -0,0 +1,254 @@
+		return -EINVAL;
+	}
+
+	if (w1_reset_bus(cfg->bus) <= 0) {
+		LOG_ERR("No 1-Wire devices connected");
+		return -EINVAL;
+	}
```
Shouldn't here return value be `-ENODEV`, that would be in line with return value from `w1_read_rom` that returns `-ENODEV` in exactly the same, bus reset, case.

str4t0m replied:
I do agree that this should be in line with the others and have applied the suggestion.
Furthermore, I have moved this check to the beginning of ds18b20_configure.
Additionally, I have changed the locking in temperature_convert() and read_power_supply().

de-nordic commented on drivers/w1/w1_net.c:
```diff
@@ -0,0 +1,405 @@
+		} else {
+			*(uint64_t *)&found_rom = found_rom_inv_64;
+		}
+		LOG_DBG("rom-found: nr.%u, %016llx", found_cnt,
```
`ROM ID-found`, if `nr` is contraction of `number` then  no dot after it.

de-nordic commented on drivers/w1/w1_net.c:
```diff
@@ -0,0 +1,405 @@
+
+	ret = reset_select(dev, config);
+	if (ret != 0) {
+		goto out;
```
There are no locks in this function, no point in doing the `goto out`, will cost you as much as just `return` from here.

de-nordic reviewed (commented):
Left some comments.

mbolivar-nordic commented on doc/reference/peripherals/w1.rst:
```diff
@@ -0,0 +1,94 @@
+.. _w1_api:
+
+W1: Dallas 1-Wire Interface
```
An entry in doc/reference/api/overview.rst is also necessary.

mbolivar-nordic commented on dts/bindings/w1/w1-device.yaml:
```diff
@@ -0,0 +1,23 @@
+          communication.
+          The driver will automatically use OVERDRIVE_SKIP and OVERDRIVE_MATCH
+          command instead of the standard commands.
+    label:
+      required: true
```
Why is this required? We shouldn't need to require labels anymore.

mbolivar-nordic commented on drivers/sensor/ds18b20/ds18b20.c:
```diff
@@ -0,0 +1,259 @@
+#define DS18B20_CONFIG_INIT(inst)					\
+	{								\
+		.bus = DEVICE_DT_GET(DT_INST_BUS(inst)),		\
+		.family = (uint8_t)DT_INST_PROP(inst, family_code),	\
```
You cannot safely use DT_INST_PROP on a property without `required: true`. Should this be using `DT_INST_PROP_OR` instead? Or maybe the maxim,dsb18b20.yaml file is missing a `required: true` for this property?

mbolivar-nordic commented on dts/bindings/w1/w1-device.yaml:
```diff
@@ -0,0 +1,23 @@
+      type : int
+      description: |
+          8-bit 1-Wire family code, which is also part of the 64 bit ROM ID.
+    overdrive-speed:
```
This doesn't seem to be used anywhere.

mbolivar-nordic commented on include/drivers/w1.h:
```diff
@@ -0,0 +1,330 @@
+}
+
+/**
+ * @brief Configure parameters of W1-Master.
```
Commit log says:

```
    Instead of master/slave terminology controller/peripheral terms are
    used to refer to bus participants.
```

But 'master' and 'slave' appear many times in this PR. What's going on?

mbolivar-nordic reviewed (dismissed):
Nice effort! I have a few requests and questions, mainly on the devicetree side.

KondratevAndrei commented on drivers/sensor/ds18b20/ds18b20.c:
```diff
@@ -0,0 +1,259 @@
+	uint16_t temp = sys_get_le16(temp_raw);
+
+	val->val1 = (temp & 0x0FFF) >> 4;
+	val->val2 = (temp & 0x000F) * 10000;
```
I think this is wrong.  My variant of calculating the fractional part of the temperature: 
`val->val2 = ((temp & 0x000F) * 1000000) >> 4;`

hubpav replied:
@KondratevAndrei Thanks for reporting this. We came across the same issue while using it.

rdrill replied:
Since the value in **scratchpad.temp** is stored as a signed 16 bit integer and **sys_get_le16()** returns unsigned value, we have to do this:
`int16_t temp = (int16_t)sys_get_le16(temp_raw);`
in other way, the tempere values below 0 were wrong for me.
This works well for me both with positive and negative values:
```
	int16_t temp = (int16_t)sys_get_le16(temp_raw);
	val->val1 = temp / 16;
	val->val2 = temp % 16 * 1000000 >> 4; 
```

tknapstad replied:
The existing temperature sensors in drivers/sensor (the tmpxxx family for instance) simply multiplies by a scaling factor and then do a divide by 1000000 and modulus 1000000. That will work fine here as well, there is no need for the rightshift.

```
    int16_t temp = (int16_t)sys_get_le16(temp_raw);
    int32_t microdeg = (int32_t)temp * DS18B20_TEMP_SCALE;

    val->val1 = microdeg / 1000000;
    val->val2 = microdeg % 1000000;
```

where `DS18B20_TEMP_SCALE` should be defined in ds18b20.h:
```
#define DS18B20_TEMP_SCALE (1000000/16)
```

hubpav commented:
The DS18B20's inherited property `reg = <0x28>;` is abused in an unfortunate manner since the most recent version (v1.6.1) of the DTC compiler produces warnings about the usage of an already once-allocated address. I suggest introducing a new DTS binding property `family` that will not be part of these sanity checks.

See the DTC changelog: https://github.com/dgibson/dtc/releases/tag/v1.6.1

hubpav commented:
> 

This is how we use multiple DS18B20 in the DTS:

```
&ds2484 {
	active-pullup;

	ds18b20_0: ds18b20_0@28 {
		compatible = "maxim,ds18b20";
		reg = <0x28>;
		resolution = <12>;
		label = "DS18B20_0";
	};

	ds18b20_1: ds18b20_1@28 {
		compatible = "maxim,ds18b20";
		reg = <0x28>;
		resolution = <12>;
		label = "DS18B20_1";
	};
};
```

dp7hgh7 commented:
Hi, what is the curent status of this PR? If necessary, I could take over fixing the details since this PR is quite importatnt for my project.

str4t0m committed:
drivers: w1 introduce 1-wire net layer api

This commit introduces a new api for the net layer of Dallas
1-wire protocol.
For single drop configurations w1_read_rom, and w1_skip_rom commands are
provided.
For multidrop configurations w1_match_rom, w1_resume_command,
w1_search_rom, and w1_search_alarm routines are provided as well.
Additionally, the reset_select routine, conditionally depending on the
bus configuration, either executes a match_rom, or a skip rom command.
A w1_write_read command simplifies the typical scenario of addressing
a device, writing a few bytes to the device and reading back the answer.

Additionally w1_crc8, and w1_crc16 are added as wrappers around the
zephyr in-tree crc8 and crc16_reflect implementations.
The former may be used to verify the validity of the rom id, while the
latter is used for integrity checking of many eeprom, and authenticator
commands.

The general search command does not support filtering on
the family code.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m reviewed (commented):
Thanks very much for all the feedback so far and sorry for the long delay.
I have now incorporated all the requested changes...

str4t0m commented:
Noteworthy additional changes since the last push:
- moved to new ztest API
- include path change to include/zephyr/...
- add optional write_block/read_block routines that drivers supporting multi-byte operations can implement.
- For consistency with the 1-Wire spec the following have been renamed:
```
w1-host.yaml			w1-master.yaml
w1-device.yaml			w1-slave.yaml

w1_get_peripheral_count		w1_get_slave_count
W1_PERIPHERALS_COUNT			W1_SLAVE_COUNT
w1_config			w1_slave_config
w1_controller_config		w1_master_config
w1_controller_data		w1_master_data
cfg→client_count		cfg→slave_count
```

de-nordic reviewed (approved):
Looks OK. Thanks!

carlescufi commented:
API meeting:

- @dp7hgh7 to check about multi-channel support before approving
- Merge afterwards

dp7hgh7 commented:
@carlescufi I rebased my work and checked with the current state of this Pr. So far I don't see any issue to go forward and merge. I'll bring up the multi channel architecture we discussed, as soon as I ported the driver.
